addons/test_mail/tests/test_mail_composer.py

```diff
@@ -0,0 +1,288 @@
+        composer.write({'template_id': self.template.id})
+        # currently onchange necessary
+        composer.onchange_template_id_wrapper()
+        self.assertEqual(composer.subject, 'About %s' % self.test_record.name)
+        self.assertEqual(composer.body, '<p>Hello %s</p>' % self.test_record.name)
+        self.assertEqual(composer.mail_server_id, self.template.mail_server_id)
+
+    @users('employee')
+    def test_mail_composer_content_mass(self):
+        composer = self.env['mail.compose.message'].with_context({
+            'default_composition_mode': 'mass_mail',
+            'default_model': self.test_record._name,
+        }).create({
+            'subject': 'My amazing subject',
+            'body': '<p>Test Body</p>',
+            'mail_server_id': False,
+            'template_id': self.template.id,
+        })
+
+        # creation values are taken
+        self.assertEqual(composer.subject, 'My amazing subject')
+        self.assertEqual(composer.body, '<p>Test Body</p>')
+        self.assertEqual(composer.mail_server_id.id, False)
+
+        # erasing template should erase its content
+        composer.write({'template_id': False})
+        # currently onchange necessary
+        composer.onchange_template_id_wrapper()
+        self.assertEqual(composer.subject, 'My amazing subject')
+        # self.assertEqual(composer.body, '<p>Test Body</p>')
+        self.assertEqual(composer.body, '')
+        self.assertEqual(composer.mail_server_id.id, False)
+
+        # changing template should update its content
+        composer.write({'template_id': self.template.id})
+        # currently onchange necessary
+        composer.onchange_template_id_wrapper()
+        self.assertEqual(composer.subject, self.template.subject)
+        self.assertEqual(composer.body, self.template.body_html)
+        self.assertEqual(composer.mail_server_id, self.template.mail_server_id)
+
+        # manual values is kept over template
+        composer.write({'subject': 'Back to my amazing subject'})
+        self.assertEqual(composer.subject, 'Back to my amazing subject')
+
+    @users('employee')
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_mail_composer_subject(self):
+        for composition_mode, records in zip(['comment', 'mass_mail'], [self.test_record, self.test_records]):
+            composer = self.env['mail.compose.message'].with_context(self._get_context(records, add_web=True)).create({})
+            self.assertEqual(composer.composition_mode, composition_mode)
+
+            # default values come from record (comment) or set to False (mass mode)
+            if composition_mode == 'comment':
+                self.assertEqual(composer.record_name, self.test_record.name)
+            else:
+                self.assertFalse(composer.record_name)
+            if composition_mode == 'comment':
+                self.assertEqual(composer.subject, 'Re: %s' % self.test_record.name)
```

Comment:
Why not group your assertions since the if/else is the same?

Reply:
Cleaned.